client/components/sites-dropdown/index.jsx

```diff
@@ -34,39 +36,35 @@ export default class SitesDropdown extends PureComponent {
 		isPlaceholder: false
 	}
 
-	constructor( props ) {
-		super( props );
-
-		this.selectSite = this.selectSite.bind( this );
-		this.toggleOpen = this.toggleOpen.bind( this );
-		this.onClose = this.onClose.bind( this );
-
-		const selectedSite = props.selectedSiteId
-			? sites.getSite( props.selectedSiteId )
-			: sites.getPrimary();
-
-		this.state = {
-			selectedSiteSlug: selectedSite && selectedSite.slug
-		};
+	state = {
+		selectedSiteId: this.props.selectedOrPrimarySiteId,
 	}
 
-	getSelectedSite() {
-		return sites.getSite( this.state.selectedSiteSlug );
-	}
+	selectSite = ( siteOrsiteID ) => {
+		const ID = Number.isInteger( siteOrsiteID )
+			? siteOrsiteID
+			: get( siteOrsiteID, 'ID', null );
```

Comment:
Hmm, I don't get the point of this, and generally, I'd like to avoid introducing even more degrees of freedom since they normally will lead to weird usage that's even harder to refactor/maintain.

This is a callback prop we pass down to `SiteSelector`. If we leave that component unchanged, it will continue to call this method with a site slug, so we don't need any logic to distinguish cases.

To allow consumers of `SitesDropdown` to have the `onSiteSelect` prop _they_ pass to `SitesDropdown` get a site ID (instead of a slug), we just need to do that; and we should also be able to use the ID for internal state here.